nntrainer/compiler/compiler.h

```diff
@@ -97,12 +113,124 @@ class GraphInterpreter {
+ * @brief Pure Virtual class for ExecutableGraph
+ *
+ * @note this class is an interface to executable graph as actual graph would be
+ * different framework by framework
+ *
+ */
+class ExecutableGraph {
+  /// NYI!
+};
 
 /**
- * @brief a simple scaffoliding function
+ * @brief Graph Representation
+ *
+ * @note this class must have the graph representation of the most expanded
+ * version as far as nntrainer supports.
+ *
+ * GraphRepresentation contains the operation configurations, connections and
+ * buffer(this might not...)
  *
  */
-void hello_world();
+class GraphRepresentation {
+public:
+  /**
+   * @brief Construct a new Graph Representation object
+   *
+   * @param interpreter_ interpreter to use
+   * @param compiler_ compiler to use
+   */
+  GraphRepresentation(std::shared_ptr<GraphInterpreter> interpreter_ = nullptr,
+                      std::shared_ptr<GraphCompiler> compiler_ = nullptr) :
+    interpreter(interpreter_),
+    compiler(compiler_) {}
+
+  /**
+   * @brief Construct a new Graph Representation object from ifstream
+   *
+   * @param interpreter_ interpreter to use
+   * @param istream istream to convert to the representation
+   */
+  GraphRepresentation(std::shared_ptr<GraphInterpreter> interpreter_,
+                      std::istream &in) {
+    /** NYI! */
+  }
+
+  /**
+   * @brief Construct a new Graph Representation object from a executable graph
+   *
+   * @param compiler_ compiler to use
+   * @param graph graph that will be decompiled (not owing)
+   */
+  GraphRepresentation(std::shared_ptr<GraphCompiler> compiler_,
+                      std::shared_ptr<ExecutableGraph> graph) {
+    /** NYI! */
+  }
+
+  /**
+   * @brief compile a graph to a executable graph
+   *
+   * @return std::shared_ptr<ExecutableGraph>
+   */
+  std::shared_ptr<ExecutableGraph> compile() {
+    /** NYI !*/
+    return nullptr;
+  };
+
+  /**
+   * @brief decompile the graph to a representation
+   *
+   * @param graph graph to decompile
+   */
+  void decompile(std::shared_ptr<ExecutableGraph> graph){/** NYI! */};
+
+  /**
+   * @brief out operator to a file
+   *
+   * @param out outstream
+   * @return std::ostream& outstream
+   */
+  std::ostream &operator<<(std::ostream &out) {
+    interpreter->serialize(*this, out);
+    return out;
+  }
+
+  /**
+   * @brief in stream
+   *
+   * @param input
+   * @return std::istream&
+   */
+  std::istream &operator>>(std::istream &input) {
+    interpreter->deserialize(input);
+    return input;
+  }
+
+  /**
+   * @brief Set the Graph Interpreter object
+   *
+   * @param interpreter_ interpreter
+   */
+  void setGraphInterpreter(std::shared_ptr<GraphInterpreter> interpreter_) {
+    interpreter = interpreter_;
+  }
+
+  /**
+   * @brief Set the Graph Compiler object
+   *
+   * @param compiler_ compiler
+   */
+  void setGraphCompiler(std::shared_ptr<GraphCompiler> compiler_) {
+    compiler = compiler_;
+  }
+
+private:
+  std::shared_ptr<GraphInterpreter> interpreter;
```

Comment:
I think GraphInterpreter and Compiler has graphrepresentation. Could you share your thought?

Reply:
Sorry for the messy commit from waiting PR, now this commit is rebased :)